cpp/src/arrow/compute/function_internal.h

```diff
@@ -0,0 +1,565 @@
+  static Result<std::vector<T>> FromScalar(const std::shared_ptr<Scalar>& value) {
+    if (value->type->id() != Type::LIST) {
+      return Status::Invalid("Expected type LIST but got ", value->type->ToString());
+    }
+    const auto& holder = checked_cast<const BaseListScalar&>(*value);
+    if (!holder.is_valid) return Status::Invalid("Got null scalar");
+    std::vector<T> result;
+    for (int i = 0; i < holder.value->length(); i++) {
+      ARROW_ASSIGN_OR_RAISE(auto scalar, holder.value->GetScalar(i));
+      ARROW_ASSIGN_OR_RAISE(auto v, PropertyFromScalar<T>::FromScalar(scalar));
+      result.push_back(std::move(v));
+    }
+    return result;
+  }
+};
+
+// Helper to quickly define serialization for a C-style enum
+#define MAKE_ENUM_PROPERTY(ENUM, TYPE, MIN, MAX)                                        \
+  template <>                                                                           \
+  struct PropertyToScalar<ENUM> {                                                       \
+    static Result<std::shared_ptr<Scalar>> ToScalar(const ENUM value) {                 \
+      return PropertyToScalar<TYPE>::ToScalar(static_cast<TYPE>(value));                \
+    }                                                                                   \
+    static Result<std::shared_ptr<DataType>> GetTypeSingleton() {                       \
+      return PropertyToScalar<TYPE>::GetTypeSingleton();                                \
+    }                                                                                   \
+  };                                                                                    \
+  template <>                                                                           \
+  struct PropertyFromScalar<ENUM> {                                                     \
+    static Result<ENUM> FromScalar(const std::shared_ptr<Scalar>& value) {              \
+      ARROW_ASSIGN_OR_RAISE(auto raw_val, PropertyFromScalar<TYPE>::FromScalar(value)); \
+      if (raw_val < static_cast<TYPE>(MIN) || raw_val > static_cast<TYPE>(MAX)) {       \
+        return Status::Invalid("Got invalid value for ", ARROW_STRINGIFY(ENUM), ": ",   \
+                               raw_val);                                                \
+      }                                                                                 \
+      return static_cast<ENUM>(raw_val);                                                \
+    }                                                                                   \
+  };
+
+#define MAKE_OPTIONS_TYPE(OPTIONS, CLASS, GETTER, PROPERTIES)                   \
+  using CLASS = decltype(PROPERTIES)::Apply<GenericOptionsType, OPTIONS>::type; \
+  const FunctionOptionsType* GETTER() {                                         \
+    static const CLASS instance(PROPERTIES);                                    \
+    return &instance;                                                           \
+  }
+
```

Comment:
```suggestion
template <typename Options, typename... Properties>
const FunctionOptionsType* GetFunctionOptionsType(const Properties&... properties) {
  static const struct : GenericOptionsType<Options, Properties...> {
    using GenericOptionsType<Options, Properties>::GenericOptionsType;
  } instance(arrow::internal::MakeProperties(properties...));

  return &instance;
}
```

(It'd probably be fine to inline `GenericOptionsType` there, actually.) Then this can be used later like so:

```c++
static auto kCastOptionsType = GetFunctionOptionsType<CastOptions>(
    arrow::internal::DataMember("to_type", &CastOptions::to_type),
    arrow::internal::DataMember("allow_int_overflow", &CastOptions::allow_int_overflow),
    /*...*/);
}  // namespace

void RegisterScalarCast(FunctionRegistry* registry) {
  DCHECK_OK(registry->AddFunction(std::make_shared<CastMetaFunction>()));
  DCHECK_OK(registry->AddFunctionOptionsType(kCastOptionsType));
}
}  // namespace internal

CastOptions::CastOptions(bool safe)
    : FunctionOptions(kCastOptionsType),
      allow_int_overflow(!safe),
      //...
```